keps/sig-windows/20190714-windows-csi-support.md

```diff
@@ -0,0 +1,401 @@
+5. Build CSI Node Plugin container images based on Windows base images.
+6. Create DaemonSet YAMLs referring to CSI Node Driver Registrar container images [3] and CSI Node Plugin container images [5] targeting Windows.
+
+Besides the above enhancements, a new "privileged proxy" binary, named csi-proxy.exe is a key aspect of this KEP. csi-proxy.exe will run as a native Windows process on the Windows nodes configured as a Windows Service. csi-proxy.exe will expose an API (through GRPC over a named pipe) for executing privileged storage related operations on Windows hosts on behalf of Windows containers like CSI Node Plugins.
+
+### User Stories
+
+With the KEP implemented, administrators should be able to deploy CSI Node Plugins that support staging, publishing and other storage management operations on Windows nodes. Operators will be able to schedule Windows workloads that consume persistent storage surfaced by CSI plugins on Windows nodes.
+
+#### Deploy CSI Node Plugin DaemonSet targeting Windows nodes 
+
+An administrator should be able to deploy a CSI Node Plugin along with the CSI Node Driver Registrar container targeting Windows nodes with a DaemonSet YAML like the following:
+
+```
+kind: DaemonSet
+apiVersion: apps/v1
+metadata:
+  name: csi-gce-pd-node-win
+spec:
+  selector:
+    matchLabels:
+      app: gcp-compute-persistent-disk-csi-driver-win
+  template:
+    metadata:
+      labels:
+        app: gcp-compute-persistent-disk-csi-driver-win
+    spec:
+      serviceAccountName: csi-node-sa
+      tolerations:
+      - key: "node.kubernetes.io/os"
+        operator: "Equal"
+        value: "win1809"
+        effect: "NoSchedule"
+      nodeSelector:
+        beta.kubernetes.io/os: windows
+      containers:
+        - name: csi-driver-registrar
+          image: gke.gcr.io/csi-node-driver-registrar:win-v1 
+          args:
+            - "--v=5"
+            - "--csi-address=unix://C:\\csi\\csi.sock"
+            - "--kubelet-registration-path=C:\\var\\lib\\kubelet\\plugins\\pd.csi.storage.gke.io\\csi.sock"
+          env:
+            - name: KUBE_NODE_NAME
+              valueFrom:
+                fieldRef:
+                  fieldPath: spec.nodeName
+          volumeMounts:
+            - name: plugin-dir
+              mountPath: C:\csi
+            - name: registration-dir
+              mountPath: C:\registration
+        - name: gce-pd-driver
+          image: gke.gcr.io/gcp-compute-persistent-disk-csi-driver:win-v1 
+          args:
+            - "--v=5"
+            - "--endpoint=unix:/csi/csi.sock"
+          volumeMounts:
+            - name: kubelet-dir
+              mountPath: C:\var\lib\kubelet
+            - name: plugin-dir
+              mountPath: C:\csi
+            - name: csi-proxy-pipe
+              mountPath: \\.\pipe\csi-proxy
+      volumes:
+        - name: csi-proxy-pipe
+          hostPath: 
+            path: \\.\pipe\csi-proxy
```

Comment:
shouldn't this be `unix:/csi/csi.sock` ?

Reply:
This is the named pipe exposed by csi-proxy.exe to the CSI node plugins. In the path between CSI node plugins and csi-proxy.exe, csi-proxy.exe is the host based process that listens to RPC calls and services them. That scenario appears to always fail in Windows when using domain sockets with permission denied errors. It does work when a host/containerized process invokes RPCs on a process in a container which is the kubelet.exe/CSI node driver registrar => CSI node plugin.